Libplanet/Blockchain/BlockChainStates.cs

```diff
@@ -56,85 +62,112 @@ public BlockChainStates(IStore store, IStateStore stateStore)
+                ? FungibleAssetValue.FromRawValue(currency, i)
+                : currency * 0;
         }
 
         /// <inheritdoc cref="IBlockChainStates.GetValidatorSet"/>
-        public ValidatorSet GetValidatorSet(BlockHash offset)
+        public ValidatorSet GetValidatorSet(BlockHash? offset)
+        {
+            ITrie stateRoot = GetStateRoot(offset);
+            string[] stringKeys = new[] { ValidatorSetKey };
+            IReadOnlyList<IValue?> rawValues = stateRoot.Get(
+                stringKeys.Select(StateStoreExtensions.EncodeKey).ToList());
+            return rawValues.Count > 0 && rawValues[0] is List list
+                ? new ValidatorSet(list)
+                : new ValidatorSet();
+        }
+
+        /// <summary>
+        /// Returns the state root associated with <see cref="BlockHash"/>
+        /// <paramref name="offset"/>.
+        /// </summary>
+        /// <param name="offset">The <see cref="BlockHash"/> to look up in
+        /// the internally held <see cref="IStore"/>.</param>
+        /// <returns>An <see cref="ITrie"/> representing the state root associated with
+        /// <paramref name="offset"/>.</returns>
+        /// <exception cref="ArgumentException">Thrown for one of the following reasons.
+        /// <list type="bullet">
+        ///     <item><description>
+        ///         If <paramref name="offset"/> is not <see langword="null"/> and
+        ///         <paramref name="offset"/> cannot be found in <see cref="IStore"/>.
+        ///     </description></item>
+        ///     <item><description>
+        ///         If <paramref name="offset"/> is not <see langword="null"/> and
+        ///         the state root hash associated with <paramref name="offset"/>
+        ///         cannot be found in <see cref="IStateStore"/>.
+        ///     </description></item>
+        /// </list>
+        /// </exception>
+        internal ITrie GetStateRoot(BlockHash? offset)
         {
-            HashDigest<SHA256>? stateRootHash = _store.GetStateRootHash(offset);
-            if (stateRootHash is { } h && _stateStore.ContainsStateRoot(h))
+            if (!(offset is { } hash))
             {
-                IReadOnlyList<IValue?> values =
-                    _stateStore.GetStates(stateRootHash, new[] { ValidatorSetKey });
-                return values.Count > 0 && values[0] is List l
-                    ? new ValidatorSet(l)
-                    : new ValidatorSet();
+                return _stateStore.GetStateRoot(null);
+            }
+            else if (_store.GetStateRootHash(hash) is { } stateRootHash)
+            {
+                if (_stateStore.ContainsStateRoot(stateRootHash))
+                {
+                    return _stateStore.GetStateRoot(stateRootHash);
+                }
+                else
+                {
+                    throw new ArgumentException(
+                        $"Could not find state root {stateRootHash} in {nameof(IStateStore)}.");
```

Comment:
adding `offset` to the exception might be good for investigation.

Reply:
Suggestion applied. 🙂